Lab 1: Creating an API Contract for the Intergalactic Marketplace

ikeepcalm opened the pull request:
Progress report:


- [x] API Contract for Product Entity (CRUD) @ikeepcalm 
- [x] OpenAPI documentation @ikeepcalm 
- [x] Info about DTOs fields, validation requirements @ikeepcalm 
- [x] Global Exceptions Handler & ControllerAdvice @Djecka-2019 
- [x] Basic domain entities @ikeepcalm 
- [x] Basic DTOs for entities @ikeepcalm 
- [x] Entity <-> DTO mapping via MapStruct @ikeepcalm 
- [x] CRUD operations for Product entity @Djecka-2019 
- [x] Validation, validations, custom validator @Djecka-2019 

Read more in [README.md](README.md)

restClient (docker + wiremock) was **NOT** implemented; simple mock was data used instead (as well as H2 database for the app to start up and mock Product repository, to be replaced with actual JpaRepository further on) @ikeepcalm

ikeepcalm committed:
Main domain entities

ikeepcalm committed:
DTO classes

ikeepcalm committed:
Contract, MapStruct, Requests

ikeepcalm committed:
Create README.md

Djecka-2019 committed:
implement ProductService

Djecka-2019 committed:
implement Exception handling with RFC 9457 standart

Djecka-2019 committed:
implement data validation, add custom validation, so names must contain cosmic words

ikeepcalm committed:
Fix MapStruct with Lombok; SecurityConfig; Swagger API; README.md with project progress

Ekatereana commented on build.gradle:
```diff
@@ -19,6 +19,15 @@ repositories {
 }
 
 dependencies {
+    runtimeOnly 'org.mariadb.jdbc:mariadb-java-client'
```
okeeeey

Ekatereana commented on build.gradle:
```diff
@@ -19,6 +19,15 @@ repositories {
 
 dependencies {
+    runtimeOnly 'org.mariadb.jdbc:mariadb-java-client'
+    runtimeOnly 'com.h2database:h2'
```
that's.... interesting 
<img width="220" height="198" alt="image" src="https://github.com/user-attachments/assets/cfa551bb-6576-480a-b5d4-22e546585dc1" />

Ekatereana commented on src/main/java/net/cosmocat/marketplace/config/OpenApiConfig.java:
```diff
@@ -0,0 +1,41 @@
+    @Bean
+    public OpenAPI cosmocatMarketplaceOpenAPI() {
+        Server devServer = new Server();
+        devServer.setUrl("http://localhost:8080");
```
That's cool, but you can do it better. Use https://springdoc.org/#swagger-ui-properties - u can do all the configurations on the level of the application.yml

ikeepcalm replied:
Isn't it only for things like path, ui-path, layout, etc (which we already have in application.properties)? 

```properties
springdoc.api-docs.path=/api-docs
springdoc.swagger-ui.path=/swagger-ui
springdoc.swagger-ui.operations-sorter=method
springdoc.swagger-ui.tags-sorter=alpha
```

For setting up meta-info such as Server, License, Name, etc - isn't it the only way to use OpenAPI or GroupedOpenAPI bean programmatically? 

https://springdoc.org/#can-i-customize-openapi-object-programmatically

Ekatereana commented on src/main/resources/api-specs/api-docs.json:
```diff
@@ -0,0 +1 @@
+{"openapi":"3.1.0","info":{"title":"Cosmocat Marketplace API","description":"This API exposes endpoints to manage products in the Cosmocat Marketplace.","contact":{"name":"Cosmocat Marketplace Team","email":"cosmocat@marketplace.com"},"license":{"name":"MIT License","url":"https://choosealicense.com/licenses/mit/"},"version":"1.0"},"servers":[{"url":"http://localhost:8080","description":"Server U ...
```
oh no... oh noooooo. my dear friends this should be readable

Ekatereana commented on src/main/java/net/cosmocat/marketplace/mapper/CategoryMapper.java:
```diff
@@ -0,0 +1,13 @@
+@Mapper(componentModel = "spring")
+public interface CategoryMapper {
+
+    CategoryDTO toDTO(Category category);
```
to what DTO? toCategoryDto

Ekatereana commented on src/main/java/net/cosmocat/marketplace/database/entity/OrderItem.java:
```diff
@@ -0,0 +1,33 @@
+
+@Getter
+@Setter
+@Entity(name = "order_items")
```
Noooooo, not yet. Please clean up and leave out any repository work

Ekatereana commented on src/main/java/net/cosmocat/marketplace/database/dto/response/ErrorResponse.java:
```diff
@@ -0,0 +1,56 @@
+@Builder
+@JsonInclude(JsonInclude.Include.NON_NULL)
+@Schema(description = "RFC 9457 compliant error response")
+public class ErrorResponse {
```
Aha, catched you - we have ProblemDetail class - the class that I mentioned in my lecture 
<img width="500" height="500" alt="image" src="https://github.com/user-attachments/assets/d79fe3ac-6c7f-48ef-900a-1778a7aa940e" />

Ekatereana commented on src/main/java/net/cosmocat/marketplace/database/dto/entity/ProductDTO.java:
```diff
@@ -0,0 +1,28 @@
+
+import java.time.LocalDateTime;
+
+@Data
```
why @data? data goest against the principle of the immutability for the dto objects :)

Ekatereana commented on src/main/java/net/cosmocat/marketplace/database/dto/request/CategoryCreateRequest.java:
```diff
@@ -0,0 +1,28 @@
+
+@Data
+@Schema(description = "Request object for creating a new category")
+public class CategoryCreateRequest {
```
You have inconcistency - Dto, request... what is the difference? Better to stick to single terminology. Dto should it be

Ekatereana commented on src/main/java/net/cosmocat/marketplace/database/dto/request/CategoryCreateRequest.java:
```diff
@@ -0,0 +1,28 @@
+
+import java.util.List;
+
+@Data
```
again - data - for mutable objects and @Value is for immutable

ikeepcalm committed:
Introduce fixes to comply with PR review comments

ikeepcalm commented:
Tried fixing all of the comments above:

- [x] Removed MariaDB and H2 dependencies
> I tested if the application would launch at all, and It had complaints about no Data-source configuration (and I was lazy to create local database for that), so I decided to just add H2 to the class-path. Since we do not use any of repositories anymore, we can remove that, and it will work just as it is.

- [x] Removed all repositories stuff, most of the entities :( 
> We will return them back anyway in the lab-2, using git commit history toolkit, I suppose 😈

- [x] Mutable @Data -> Immutable @Value for DTO classes
- [x] Fixed inconsistencies in DTO Request namings
- [x] Mapper's methods' names now match the mapped entity
- [x] Custom ErrorResponse was replaced with Spring's ProblemDetail
- [x] Ctrl + Alt + L for the api-specs/api-docs.json to make it readable

- [ ] Couldn't solve the comment regarding OpenApiConfig yet

ikeepcalm committed:
Formatted the code according to Google Java style

Ekatereana commented on src/main/java/net/cosmocat/marketplace/controller/ProductController.java:
```diff
@@ -0,0 +1,185 @@
+            @PathVariable Long id,
+            @Valid @RequestBody ProductUpdateDTO request) {
+        log.info("Updating product with ID: {}", id);
+        Optional<ProductDTO> updatedProduct = productService.updateProduct(id, request);
```
Ops, ok so the problem here - I love how you use optional that's awesome, but: it's bad practice to return the optional from the methods. Why - because optional in it's nature could have 2 states, and you making the user of your method responsible for handling it. So it breaks separation of concern principle. So it's never good idea to return the optional from the methods

Ekatereana replied:
second problem - you break the spring flow of handling the response -> basically you should just throw the expection on the level of the service. This will be handled by ur controller advice, and there you will return 404, because it's exactly how single source of truth handling works. Here you will just return resonse Entiy ok always

Ekatereana commented on src/main/java/net/cosmocat/marketplace/controller/ProductController.java:
```diff
@@ -0,0 +1,185 @@
+                    responseCode = "204",
+                    description = "Product deleted successfully"
+            ),
+            @ApiResponse(
```
404 - Delete is idempotent, please re-watch the lecture
<img width="540" height="500" alt="image" src="https://github.com/user-attachments/assets/4e2c6815-cc8e-4c5b-b7c4-b3deb8fe3ef3" />

Ekatereana commented on src/main/java/net/cosmocat/marketplace/controller/ProductController.java:
```diff
@@ -0,0 +1,185 @@
+        log.info("Deleting product with ID: {}", id);
+        boolean deleted = productService.deleteProduct(id);
+
+        return deleted ? ResponseEntity.noContent().build()
```
again same, should be all handled in controller advice

Ekatereana commented on src/main/java/net/cosmocat/marketplace/database/dal/service/ProductService.java:
```diff
@@ -0,0 +1,189 @@
+  }
+
+  public List<ProductDTO> getAllProducts() {
+    log.debug("Retrieving all products");
```
usually that's info logs, you wanna see them when the application is run normally

Ekatereana commented on src/main/java/net/cosmocat/marketplace/database/dal/service/ProductService.java:
```diff
@@ -0,0 +1,189 @@
+    return productMapper.toDTOList(productList);
+  }
+
+  public Optional<ProductDTO> getProductById(Long id) {
```
Optional of nullable is good, bad to return it in the method, and here you should throw the exeption

Ekatereana commented on src/main/java/net/cosmocat/marketplace/database/dto/response/CosmoApiResponse.java:
```diff
@@ -0,0 +1,39 @@
+@AllArgsConstructor
+@NoArgsConstructor
+@Schema(description = "Standard API response wrapper")
+public class CosmoApiResponse<T> {
```
In general I don't see why you need this object, Response entity that's what doing the wrapping functionality for you. It's not mistake, just don't see why it's needed so that's why I'm questioning

Ekatereana commented on src/main/java/net/cosmocat/marketplace/exception/type/ResourceNotFoundException.java:
```diff
@@ -0,0 +1,23 @@
+
+public class ResourceNotFoundException extends RuntimeException {
+
+  public ResourceNotFoundException(String message) {
```
better you specialized exeptions: Product, order, category not found. It would be cleaner and more consistent

Ekatereana commented on src/main/java/net/cosmocat/marketplace/validation/CosmicWordCheckValidator.java:
```diff
@@ -0,0 +1,99 @@
+  }
+
+  @Override
+  public boolean isValid(String value, ConstraintValidatorContext context) {
```
I love how you do it in the own way but you can do it much easier with the pattern mattcher :)

Ekatereana commented on src/main/resources/api-specs/api-docs.json:
```diff
@@ -0,0 +1,515 @@
+              }
+            }
+          },
+          "404": {
```
nope.

Ekatereana commented on src/main/resources/api-specs/api-docs.json:
```diff
@@ -0,0 +1,515 @@
+              }
+            }
+          },
+          "201": {
```
where is conflict? would be nice to have

Djecka-2019 committed:
Refactor ProductController and ProductService to remove Optional usage

Djecka-2019 committed:
Make delete always return 204

Djecka-2019 committed:
Change log debug to log info

Djecka-2019 committed:
Implement resource conflict handling for duplicate

Djecka-2019 committed:
Specialise exception handling for product and category

Djecka-2019 committed:
change CosmicWordCheckValidator to use pattern matcher

ikeepcalm committed:
Remove redundant CosmoApiResponse, replace with ResponseEntity directly

ikeepcalm committed:
Regenerate api-docs.json according to fixes

Ekatereana commented on src/main/java/net/cosmocat/marketplace/controller/ProductController.java:
```diff
@@ -0,0 +1,170 @@
+                    content = @Content(schema = @Schema(implementation = ProductDTO.class))
+            )
+    })
+    @GetMapping("/search")
```
I like this. good that you do a bit extra

Ekatereana commented on src/main/java/net/cosmocat/marketplace/database/dto/entity/CategoryDTO.java:
```diff
@@ -0,0 +1,13 @@
+import java.util.List;
+import lombok.Value;
+
+@Value
```
GOOOOD!

Ekatereana commented on src/main/java/net/cosmocat/marketplace/database/dto/request/ProductUpdateDTO.java:
```diff
@@ -0,0 +1,76 @@
+public class ProductUpdateDTO {
+
+  @Size(min = 2, max = 100, message = "Product name must be between 2 and 100 characters")
+  @CosmicWordCheck(
```
Good + 1
